src/main/java/winstone/Launcher.java

```diff
@@ -229,8 +231,17 @@ private void writePortToFileIfNeeded() throws IOException {
                     Path portFile = Paths.get(portFileName);
                     Path portDir = portFile.getParent();
                     Files.createDirectories(portDir);
-                    try (BufferedWriter writer = Files.newBufferedWriter(portFile, StandardCharsets.UTF_8)) {
+                    Path tmpPath = Files.createTempFile(portDir, "temp", "");
+                    try (BufferedWriter writer = Files.newBufferedWriter(tmpPath, StandardCharsets.UTF_8)) {
                         writer.write(Integer.toString(port));
+                        try {
+                            Files.move(tmpPath, portFile, StandardCopyOption.ATOMIC_MOVE);
```

Comment:
I'm a little bit confused but in very rare case `HttpConnectorFactoryTest#writePortInFile` Test could still read empty file content, does anyone knows is `StandardCopyOption.ATOMIC_MOVE` absolutely guarantee atomic move?

Reply:
what line does it fail on? line 84 to 87 are not waiting for it to exist in the test from what I can tell

Reply:
line 87, condition "Port value should not be empty in any time" (which comes from Future), fails very-very rare.

Reply:
Flakiness was because `writer` wasn't auto closed. Fixed in [d4688e8](https://github.com/jenkinsci/winstone/pull/249/commits/d4688e8c29d7d2c1067badb750f5751a87ca05b5)